// @aria-ui/elements/dist/overlay-root-MMTgTGeY.js
import { computed, createSignal, defineProps, onMount, useEffect } from "@aria-ui/core";
import { FeatureDetection, useAriaDisabled, useDataState, useDisabledMountTransitionStyle, useElementId, usePresence, useTransitionStatus } from "@aria-ui/utils";
import { autoUpdate, computePosition, flip, hide, inline, offset, shift, size } from "@floating-ui/dom";
//#region src/overlay/open-change-event.ts
var OpenChangeEvent = class extends Event {
	constructor(open) {
		super("openChange");
		this.detail = open;
	}
};
//#endregion
//#region src/overlay/overlay-popup.ts
/**
* @internal
*/
const OverlayPopupPropsDeclaration = /* @__PURE__ */ defineProps({});
/**
* @internal
*/
function setupOverlayPopup(host, getOverlayStore) {
	const id = useElementId(host);
	useEffect(host, () => {
		const store = getOverlayStore();
		if (!store) return;
		store.setPopupId(id);
	});
	const getOpen = computed(() => getOverlayStore()?.getIsOpen() ?? false);
	useDataState(host, getOpen);
	usePresence(host, getOpen);
}
//#endregion
//#region ../../node_modules/.pnpm/@ocavue+utils@1.6.0/node_modules/@ocavue/utils/dist/index.js
/**
* Checks if the given value is an object.
*/
function isObject(value) {
	return typeof value == "object" && !!value;
}
/**
* Checks if the given DOM node is an Element.
*/
function isElement(node) {
	return node.nodeType === 1;
}
/**
* Checks if the given DOM node is a Document.
*/
function isDocument(node) {
	return node.nodeType === 9;
}
/**
* Checks if the given DOM node is a DocumentFragment.
*/
function isDocumentFragment(node) {
	return node.nodeType === 11;
}
/**
* Checks if the given DOM node is a ShadowRoot.
*/
function isShadowRoot(node) {
	return isDocumentFragment(node) && "host" in node && isElementLike(node.host);
}
/**
* Checks if an unknown value is likely a DOM element.
*/
function isElementLike(value) {
	return isObject(value) && value.nodeType === 1 && typeof value.nodeName == "string";
}
/**
* Gets the window object for the given target or the global window object if no
* target is provided.
*/
function getWindow(target) {
	if (target) {
		if (isShadowRoot(target)) return getWindow(target.host);
		if (isDocument(target)) return target.defaultView || window;
		if (isElement(target)) return target.ownerDocument?.defaultView || window;
	}
	return window;
}
//#endregion
//#region src/overlay/positioning.ts
/**
* @internal
*/
function updatePlacement(floating, reference, options) {
	const middleware = [
		setupOffset(options),
		setupFlip(options),
		setupShift(options),
		setupSize(options),
		setupInline(options),
		setupTransformOrigin(options),
		setupHide(options)
	];
	const { placement, strategy } = options;
	let canceled = false;
	const update = async () => {
		if (canceled) return;
		const referenceElement = unwrapElement(reference);
		if (referenceElement && !referenceElement.isConnected) return;
		if (options.hoist && FeatureDetection.supportsPopover()) {
			floating.style.margin = "unset";
			floating.setAttribute("popover", "manual");
			floating.showPopover?.();
		}
		const pos = await computePosition(reference, floating, {
			placement,
			middleware,
			strategy
		});
		if (canceled) return;
		if (options.hide) {
			const hidden = pos.middlewareData.hide?.escaped || pos.middlewareData.hide?.referenceHidden;
			floating.style.visibility = hidden ? "hidden" : "visible";
		}
		const dpr = getDPR(floating);
		const x = Math.round(pos.x * dpr) / dpr;
		const y = Math.round(pos.y * dpr) / dpr;
		floating.style.position = pos.strategy;
		floating.style.top = "0px";
		floating.style.left = "0px";
		floating.style.transform = `translate(${x}px,${y}px)`;
		if (dpr >= 1.5) floating.style.willChange = "transform";
		const [side, align] = getSideAndAlignFromPlacement(pos.placement);
		floating.setAttribute("data-side", side);
		floating.setAttribute("data-align", align);
	};
	const autoUpdateOptions = typeof options.autoUpdate === "boolean" ? void 0 : options.autoUpdate;
	const cancelAutoUpdate = options.autoUpdate ? autoUpdate(reference, floating, update, autoUpdateOptions) : void 0;
	if (!options.autoUpdate) update();
	return () => {
		canceled = true;
		cancelAutoUpdate?.();
	};
}
function getOverflowOptions(props) {
	return {
		boundary: props.boundary,
		rootBoundary: props.rootBoundary,
		elementContext: props.elementContext,
		altBoundary: props.altBoundary,
		padding: props.overflowPadding
	};
}
function setupOffset(props) {
	if (props.offset == null) return;
	return offset(props.offset);
}
function setupFlip(props) {
	if (!props.flip) return;
	return flip({
		...getOverflowOptions(props),
		fallbackPlacements: props.flip === true ? void 0 : props.flip,
		crossAxis: false
	});
}
function setupShift(props) {
	if (!props.shift) return;
	return shift({
		...getOverflowOptions(props),
		mainAxis: props.shift,
		crossAxis: props.overlap
	});
}
function setupSize(props) {
	if (!props.fitViewport && !props.sameWidth && !props.sameHeight) return;
	return size({
		...getOverflowOptions(props),
		apply({ elements, rects, availableHeight, availableWidth }) {
			const floating = elements.floating;
			if (props.sameWidth) floating.style.width = `${Math.round(rects.reference.width)}px`;
			if (props.sameHeight) floating.style.height = `${Math.round(rects.reference.height)}px`;
			if (props.fitViewport) {
				floating.style.maxWidth = `${Math.floor(availableWidth)}px`;
				floating.style.maxHeight = `${Math.floor(availableHeight)}px`;
			}
		}
	});
}
function setupInline(props) {
	if (!props.inline) return;
	return inline();
}
function setupTransformOrigin(props) {
	return {
		name: "transformOrigin",
		fn(state) {
			const { elements, middlewareData, placement: renderedPlacement, rects, x, y } = state;
			const [side] = getSideAndAlignFromPlacement(renderedPlacement);
			const refCenterX = rects.reference.x + rects.reference.width / 2 - x;
			const refCenterY = rects.reference.y + rects.reference.height / 2 - y;
			const transformX = Math.max(0, Math.min(refCenterX, rects.floating.width));
			const transformY = Math.max(0, Math.min(refCenterY, rects.floating.height));
			const offsetValue = getMainAxisOffset(props.offset);
			const isOverlapping = Math.abs(middlewareData.shift?.y || 0) > offsetValue;
			const sideAxis = side === "top" || side === "bottom" ? "y" : "x";
			let transformOrigin;
			if (props.overlap && sideAxis === "y" && isOverlapping) {
				const halfAnchorHeight = rects.reference.height / 2;
				transformOrigin = `${transformX}px ${rects.reference.y + halfAnchorHeight - y}px`;
			} else transformOrigin = {
				top: `${transformX}px calc(100% + ${offsetValue}px)`,
				bottom: `${transformX}px ${-offsetValue}px`,
				left: `calc(100% + ${offsetValue}px) ${transformY}px`,
				right: `${-offsetValue}px ${transformY}px`
			}[side];
			elements.floating.style.setProperty("--transform-origin", transformOrigin);
			return {};
		}
	};
}
function getMainAxisOffset(offset) {
	if (typeof offset === "number") return offset;
	if (typeof offset === "object" && offset !== null) return offset.mainAxis ?? 0;
	return 0;
}
function setupHide(props) {
	if (!props.hide) return;
	return hide({
		padding: props.overflowPadding,
		elementContext: "reference"
	});
}
function getDPR(element) {
	return getWindow(element).devicePixelRatio || 1;
}
function getSideAndAlignFromPlacement(placement) {
	const [side, align = "center"] = placement.split("-");
	return [side, align];
}
/**
* Unwraps a virtual element to its underlying DOM element.
*
* Copied from https://github.com/floating-ui/floating-ui/blob/b80ccaf9cfd7f80f28546906c60284c8385940f0/packages/dom/src/utils/unwrapElement.ts
*
* @internal
*/
function unwrapElement(element) {
	return !isElementLike(element) ? element.contextElement : element;
}
//#endregion
//#region src/overlay/overlay-positioner.ts
/**
* @internal
*/
const OverlayPositionerPropsDeclaration = /* @__PURE__ */ defineProps({
	strategy: {
		default: "absolute",
		attribute: "strategy",
		type: "string"
	},
	placement: {
		default: "top",
		attribute: "placement",
		type: "string"
	},
	autoUpdate: {
		default: true,
		attribute: false
	},
	hoist: {
		default: true,
		attribute: "hoist",
		type: "boolean"
	},
	offset: {
		default: 6,
		attribute: false
	},
	flip: {
		default: true,
		attribute: false
	},
	shift: {
		default: true,
		attribute: "shift",
		type: "boolean"
	},
	overlap: {
		default: false,
		attribute: "overlap",
		type: "boolean"
	},
	fitViewport: {
		default: false,
		attribute: "fit-viewport",
		type: "boolean"
	},
	sameWidth: {
		default: false,
		attribute: "same-width",
		type: "boolean"
	},
	sameHeight: {
		default: false,
		attribute: "same-height",
		type: "boolean"
	},
	inline: {
		default: false,
		attribute: "inline",
		type: "boolean"
	},
	hide: {
		default: false,
		attribute: "hide",
		type: "boolean"
	},
	boundary: {
		default: "clippingAncestors",
		attribute: false
	},
	rootBoundary: {
		default: "viewport",
		attribute: "root-boundary",
		type: "string"
	},
	overflowPadding: {
		default: 4,
		attribute: "overflow-padding",
		type: "number"
	},
	elementContext: {
		default: "floating",
		attribute: "element-context",
		type: "string"
	},
	altBoundary: {
		default: false,
		attribute: "alt-boundary",
		type: "boolean"
	}
});
/**
* @internal
*/
function setupOverlayPositioner(host, props, getStore) {
	const getOpen = computed(() => getStore()?.getIsOpen() ?? false);
	const getAnchorElement = computed(() => getStore()?.getAnchorElement());
	const getHoist = props.hoist.get;
	useDisabledMountTransitionStyle(host, useTransitionStatus(host, getOpen));
	useEffect(host, () => {
		const id = useElementId(host);
		getStore()?.setPositionerId(id);
	});
	let popoverDirty = false;
	if (FeatureDetection.supportsPopover()) useEffect(host, () => {
		if (getHoist()) {
			popoverDirty = true;
			host.popover = "manual";
			host.showPopover();
		} else if (popoverDirty) host.popover = null;
	});
	useDataState(host, getOpen);
	useEffect(host, () => {
		const open = getOpen();
		const anchorElement = getAnchorElement();
		if (!open || !anchorElement) return;
		return updatePlacement(host, anchorElement, {
			strategy: props.strategy.get(),
			placement: props.placement.get(),
			autoUpdate: props.autoUpdate.get(),
			hoist: props.hoist.get(),
			offset: props.offset.get(),
			flip: props.flip.get(),
			shift: props.shift.get(),
			overlap: props.overlap.get(),
			fitViewport: props.fitViewport.get(),
			sameWidth: props.sameWidth.get(),
			sameHeight: props.sameHeight.get(),
			inline: props.inline.get(),
			hide: props.hide.get(),
			boundary: props.boundary.get(),
			rootBoundary: props.rootBoundary.get(),
			overflowPadding: props.overflowPadding.get(),
			elementContext: props.elementContext.get(),
			altBoundary: props.altBoundary.get()
		});
	});
	onMount(host, () => {
		host.style.setProperty("background-color", "transparent");
	});
}
//#endregion
//#region src/overlay/overlay-store.ts
function createOverlayStore(getOpen, setOpen, getDefaultOpen, getDisabled, dispatchOpenChangeEvent) {
	const anchorElement = createSignal(void 0);
	const positionerId = createSignal("");
	const popupId = createSignal("");
	const getIsOpen = computed(() => {
		const canOpen = !getDisabled() && anchorElement.get();
		const openValue = getOpen() ?? getDefaultOpen();
		return canOpen ? openValue : false;
	});
	const requestOpenChange = (open) => {
		if (getDisabled()) return;
		const event = new OpenChangeEvent(open);
		dispatchOpenChangeEvent(event);
		if (event.defaultPrevented) return;
		setOpen(open);
	};
	const requestOpenToggle = () => {
		requestOpenChange(!getIsOpen());
	};
	return {
		getIsOpen,
		requestOpenChange,
		requestOpenToggle,
		getPositionerId: positionerId.get,
		setPositionerId: positionerId.set,
		getPopupId: popupId.get,
		setPopupId: popupId.set,
		getAnchorElement: anchorElement.get,
		setAnchorElement: anchorElement.set
	};
}
//#endregion
//#region src/overlay/overlay-root.ts
/**
* @internal
*/
const OverlayRootPropsDeclaration = defineProps({
	defaultOpen: {
		default: false,
		attribute: "default-open",
		type: "boolean"
	},
	open: {
		default: null,
		attribute: "open",
		type: "json"
	},
	disabled: {
		default: false,
		attribute: "disabled",
		type: "boolean"
	}
});
/**
* @internal
*/
function useOverlayStore(host, props) {
	const getDisabled = computed(() => props.disabled.get());
	const dispatchOpenChangeEvent = (event) => {
		host.dispatchEvent(event);
	};
	const store = createOverlayStore(props.open.get, props.open.set, props.defaultOpen.get, getDisabled, dispatchOpenChangeEvent);
	useAriaDisabled(host, getDisabled);
	return store;
}
//#endregion
export { setupOverlayPositioner as a, OpenChangeEvent as c, OverlayPositionerPropsDeclaration as i, useOverlayStore as n, OverlayPopupPropsDeclaration as o, createOverlayStore as r, setupOverlayPopup as s, OverlayRootPropsDeclaration as t };
